library/alloc/src/vec/spec_extend_with.rs

```diff
@@ -0,0 +1,37 @@
+use core::clone::TrivialClone;
+use core::mem::MaybeUninit;
+use core::{iter, ptr};
+
+use super::Vec;
+use crate::alloc::Allocator;
+
+// Specialization trait used for Vec::extend_with
+pub(super) trait SpecExtendWith<T> {
+    fn spec_extend_with(&mut self, n: usize, value: T);
+}
+
+impl<T: Clone, A: Allocator> SpecExtendWith<T> for Vec<T, A> {
+    default fn spec_extend_with(&mut self, n: usize, value: T) {
+        self.extend_trusted(iter::repeat_n(value, n));
+    }
+}
+
+impl<T: TrivialClone, A: Allocator> SpecExtendWith<T> for Vec<T, A> {
+    fn spec_extend_with(&mut self, n: usize, value: T) {
+        let len = self.len();
+        self.reserve(n);
+        let unfilled = self.spare_capacity_mut();
+
+        // SAFETY: the above `reserve` call guarantees `n` to be in bounds.
+        let unfilled = unsafe { unfilled.get_unchecked_mut(..n) };
+
+        // SAFETY: because `T` is `TrivialClone`, this is equivalent to calling
+        // `T::clone` for every element. Notably, `TrivialClone` also implies
+        // that the `clone` implementation will not panic, so we can avoid
+        // initialization guards and such.
+        unfilled.fill_with(|| MaybeUninit::new(unsafe { ptr::read(&value) }));
```

Comment:
2. This will probably optimize away, but it's a bunch of steps for something that should be quite trivial, so I wonder if that's impacting the debug build codegen speed.

What if you skip doing anything complicated and just have
```rust
for i in 0..n {
    unsafe { ptr::write(unfilled.get_unchecked_mut(i), ptr::read(&value)) };
}
```

That also means we're not copying it *as MaybeUninit*, which will give it more information since (for example) copying `MaybeUninit<&[i32]>` is different from copying `&[i32]`.